escodegen.js

```diff
@@ -1836,8 +1838,34 @@
         BinaryExpression: function (expr, precedence, flags) {
             var result, leftPrecedence, rightPrecedence, currentPrecedence, fragment, leftSource;
             currentPrecedence = BinaryPrecedence[expr.operator];
-            leftPrecedence = expr.operator === '**' ? Precedence.Postfix : currentPrecedence;
-            rightPrecedence = expr.operator === '**' ? currentPrecedence : currentPrecedence + 1;
+
+            leftPrecedence = currentPrecedence;
+            rightPrecedence = currentPrecedence + 1;
+
+            switch (expr.operator) {
+                case '**':
+                    leftPrecedence = Precedence.Postfix;
+                    rightPrecedence = currentPrecedence;
+                    break;
+
+                case '??':
+                    if (expr.left.type === Syntax.LogicalExpression && (expr.left.operator === '||' || expr.left.operator === '&&')) {
+                        leftPrecedence = BinaryPrecedence[expr.left.operator] + 1;
+                    }
+
+                    if (expr.right.type === Syntax.LogicalExpression && expr.right.operator === '&&') {
+                        rightPrecedence = BinaryPrecedence[expr.right.operator] + 1;
+                    }
+
+                    break;
+
+                case '||':
+                    if (expr.left.operator === '??') {
```

Comment:
This one too.

Reply:
Done https://github.com/estools/escodegen/pull/417/commits/6c1bd9479f02fc37ccf28008ce7e95e2de58f93a